### PR TITLE
feat: Add sample multi-agent notebook for using Reasoning Engine with LangGraph and RAG.

### PR DESCRIPTION
**Pull request by shawn-yang-google**
# Description

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [x] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [x] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [x] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by shawn-yang-google** [comment]
I'm unable to assign reviewers directly. 
Kris ([koverholt@google.com](mailto:koverholt@google.com)) is aware of this pull request, as I've reached out separately.

**Comment by koverholt** [comment]
Thanks for the PR @shawn-yang-google, reviewing this now!

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 1]
```diff
@@ -0,0 +1,1078 @@
...
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "xU0F5ObiGgF4"
+      },
+      "source": [
+        "# Deploying a RAG + LangGraph Application with Cloud SQL for Postgres with Reasoning Engine with on Vertex AI\n",
+        "\n",
+        "<table align=\"left\">\n",
+        "  <td style=\"text-align: center\">\n",
+        "    <a href=\"https://colab.research.google.com/github/GoogleCloudPlatform/generative-ai/blob/main/gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb\">\n",
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:36]: Suggest using a shorter notebook file name for simplicity, such as `tutorial_langgraph_rag_agent.ipynb`.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+        "# limitations under the License."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "xU0F5ObiGgF4"
+      },
+      "source": [
+        "# Deploying a RAG + LangGraph Application with Cloud SQL for Postgres with Reasoning Engine with on Vertex AI\n",
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:32]: Suggest reworking the title to focus more on the use case and primary tools, such as:

> Building a Multi-Agent RAG Application with LangGraph and Reasoning Engine

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 3]
```diff
@@ -0,0 +1,1078 @@
...
...AAECBAgQIECAAAECBAgQIECAAAECFwGH4wu7UAIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQINAYfjxk5aEiBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIELgIOBxf2IUSIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQaAg4HDd20pIAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIDARcDh+MIulAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgEBDwOG4sZOWBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBC4CDscXdqEECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIEGgIOx42dtCRAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBwEXA4vrALJUCAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQINAQcDhu7KQlAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAgYuAw/GFXSgBAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgACBhoDDcWMnLQkQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAhcBByOL+xCCRAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECDQEHI4bO2lJgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBA4CLgcHxhF0qAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECgIeBw3NhJSwIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIXgQHMYcmc2fnXxwAAAABJRU5ErkJggg==)\n"
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:100]: I put this image in the associated GCS bucket that we use for this repo, so you can embed an image reference to the following URL and remove the base64 version of this image:

https://storage.googleapis.com/github-repo/generative-ai/gemini/reasoning-engine/images/tutorial_langgraph_rag.png

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 4]
```diff
@@ -0,0 +1,1078 @@
...
+      "source": [
+        "!pip install --upgrade --user --quiet \\\n",
+        "    \"google-cloud-aiplatform[reasoningengine,langchain]\"==1.60.0 \\\n",
+        "    langchain-google-cloud-sql-pg \\\n",
+        "    cloud-sql-python-connector==1.9.0 \\\n",
+        "    langchain-google-vertexai \\\n",
+        "    cloudpickle==3.0.0 \\\n",
+        "    pydantic==2.7.4 \\\n",
+        "    langgraph \\\n",
+        "    httpx"
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:145]: Suggest pinning the versions of `langchain-google-cloud-sql-pg` and `langchain-google-vertexai` to the working versions in your dev environment so we can ensure this notebook + related LangChain deps are fully reproducible.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 5]
```diff
@@ -0,0 +1,1078 @@
...
+        "id": "Nqwi-5ufWp_B"
+      },
+      "outputs": [],
+      "source": [
+        "PROJECT_ID = \"reasoning-engine-test-1\"  # @param {type:\"string\"}\n",
+        "LOCATION = \"us-central1\"  # @param {type:\"string\"}\n",
+        "STAGING_BUCKET = \"gs://reasoning-engine-test-1-bucket\"  # @param {type:\"string\"}\n",
+        "\n",
+        "vertexai.init(project=PROJECT_ID, location=LOCATION, staging_bucket=STAGING_BUCKET)\n",
+        "!gcloud config set project {PROJECT_ID}"
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:274]: Change the PROJECT_ID and STAGING_BUCKET to the values in the notebook template (https://github.com/GoogleCloudPlatform/generative-ai/blob/main/notebook_template.ipynb), such as: `[your-project-id]` and `[your-bucket-name]`.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 6]
```diff
@@ -0,0 +1,1078 @@
...
+        "id": "XtiB5-LVVkv0"
+      },
+      "outputs": [],
+      "source": [
+        "REGION = \"\"  # @param {type:\"string\"}\n",
+        "INSTANCE = \"\"  # @param {type:\"string\"}\n",
+        "DATABASE = \"\"  # @param {type:\"string\"}\n",
+        "MOVIE_TABLE_NAME = \"\"  # @param {type:\"string\"}\n",
+        "BOOK_TABLE_NAME = \"\"  # @param {type:\"string\"}\n",
+        "PASSWORD = input(\"Please provide a password to be used for 'postgres' database user: \")"
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:309]: Suggest populating these with placeholder values or sample values.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 7]
```diff
@@ -0,0 +1,1078 @@
...
+        "Load data from a CSV file to generate and insert embeddings to the vector store.\n",
+        "\n",
+        "We will use two datasets:\n",
+        "\n",
+        "* Harry Potter Movie\n",
+        "  - Intro: https://www.kaggle.com/datasets/rounakbanik/the-movies-dataset\n",
+        "  - Data: gs://reasoning-engine-test-1-bucket/harry_potter/harry_potter_movies.json\n",
+        "* Harry Potter Book\n",
+        "  - Intro: https://www.kaggle.com/datasets/shubhammaindola/harry-potter-books\n",
+        "  - Data: gs://reasoning-engine-test-1-bucket/harry_potter/harry_potter_books.json\n"
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:381]: We'll need to move these sample data files into the gs://cloud-sample-data or gs://github-repo bucket for consistency and longevity with the rest of the sample data used in this repo.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 8]
```diff
@@ -0,0 +1,1078 @@
...
+        "    Returns:\n",
+        "        str: A list of BaseMessage containing the search results.\n",
+        "    \"\"\"\n",
+        "    engine = PostgresEngine.from_instance(\n",
+        "        PROJECT_ID,\n",
+        "        REGION,\n",
+        "        INSTANCE,\n",
+        "        DATABASE,\n",
+        "        quota_project=PROJECT_ID,\n",
+        "        # Uncomment to use built-in authentication instead of IAM authentication\n",
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:532]: The lines after are uncommented already. Need to update this comment to fix that and clarify what built-in authentication means (plaintext auth?).

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 9]
```diff
@@ -0,0 +1,1078 @@
...
+        "    Returns:\n",
+        "        List[BaseMessage]: A list of BaseMessage containing the search results.\n",
+        "    \"\"\"\n",
+        "    engine = PostgresEngine.from_instance(\n",
+        "        PROJECT_ID,\n",
+        "        REGION,\n",
+        "        INSTANCE,\n",
+        "        DATABASE,\n",
+        "        quota_project=PROJECT_ID,\n",
+        "        # Uncomment to use built-in authentication instead of IAM authentication\n",
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:573]: Same as previous comment.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 10]
```diff
@@ -0,0 +1,1078 @@
...
+        "def router(\n",
+        "    state: List[BaseMessage],\n",
+        ") -> Literal[\"book_similarity_search\", \"movie_similarity_search\", \"__end__\"]:\n",
+        "    if not state[0].content or len(state[1].tool_calls) == 0:\n",
+        "        return \"__end__\"\n",
+        "    if \"book\" in state[0].content:\n",
+        "        return \"book_similarity_search\"\n",
+        "    if \"movie\" in state[0].content:\n",
+        "        return \"movie_similarity_search\"\n",
+        "    return \"__end__\""
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:617]: Explain in comments or the text cell above how this router works, suggest something along the lines of: "The router will invoke different nodes in the graph based on whether the user prompt contains the word 'book' or 'movie'."

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 11]
```diff
@@ -0,0 +1,1078 @@
...
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "tZYtR6-zqudb"
+      },
+      "source": [
+        "#### Multi stage"
+      ]
+    },
+    {
+      "cel
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:779]: Suggest rendering the output of the following test cells to show users the actual output vs. expected output.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 12]
```diff
@@ -0,0 +1,1078 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BF8xqd84351O"
+      },
+      "source": [
+        "### Define router\n",
+        "\n",
+        "Then, you'll define a router to control the flow of the conversation, determining which tool to use based on user input or the state of the interaction. Here we'll use a simple router setup, and you can customize the behavior of your router to handle multiple tools, custom logic, or multi-agent workflows."
```
Comment (koverholt) [inline gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb:597]: Suggest adding a link to https://blog.langchain.dev/langgraph-multi-agent-workflows/, clarifying that we're using the `Multi Agent Collaboration` approach listed there, and mentioning that this sample notebook could be adapted to use other multi-agent implementations described on that page, such as the `Agent Supervisor` or other approaches. This is an important point you might also want to add to the Overview and/or Objective section at the top of this notebook.

**Review by koverholt — CHANGES_REQUESTED** [review]
Thanks for contributing this sample! ❤️ I appreciate the addition of another LangGraph example, the implementation of Multi-Agent Collaboration, and the explainer diagrams and text! I left some comments regarding various cleanup and tweaks, let me know if you have any questions along the way.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 1 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Changed to `tutorial_langgraph_rag_agent.ipynb`.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb [hunk 13]
```diff
@@ -0,0 +1,1078 @@
...
+        "        B{Router}\n",
+        "        C[Movie RAG]\n",
+        "        D[Book RAG]\n",
+        "        E[Summarizer]\n",
+        "    end\n",
+        "\n",
+        "    subgraph \"Output\"\n",
+        "        F[Answer]\n",
+        "    end\n",
+        "\n",
```
[the remark this answers is not in the document]
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Removed.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 12 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Added the link in both places.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 4 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Pinned the versions of all packages.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 3 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Replaced the base64 version to the URL.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 5 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Changed.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 9 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Changed.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 10 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Added.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 11 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Added the actual outputs when running the test cells.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 2 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Changed.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 7 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: I changed the urls to 
gs://github-repo/generative-ai/gemini/reasoning-engine/harry_potter/harry_potter_movies.json
and 
gs://github-repo/generative-ai/gemini/reasoning-engine/harry_potter/harry_potter_books.json

However, I don't have the access to create file in the gs://github-repo

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 6 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Added placehodler.

File: gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb (hunk 8 above)
Comment (shawn-yang-google) [reply gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb]: Removed the comment.

**Review by koverholt — APPROVED** [review]
Thanks for the updates and for contributing this notebook!

**Pull request merged**